### PR TITLE
WXDEFI-368 manager vault

### PR DESCRIPTION
**Pull request by bra1nsurfer**

### CODE REVIEW

File: ride/manager_vault.ride [hunk 1]
```diff
@@ -0,0 +1,243 @@
...
+  FOLD<20>(votes, [], clearVotesHelper)
+}
+
+func voteINTERNAL(
+    callerAddressString: String, 
+    keyPrefix: String, 
+    minVotes: Int, 
+    voteResult: 
+    List[StringEntry|IntegerEntry|DeleteEntry]
+  ) = {
```
Comment (ridev6) [inline ride/manager_vault.ride:90]:
```suggestion
  callerAddressString: String, 
  keyPrefix: String, 
  minVotes: Int, 
  voteResult: List[StringEntry|IntegerEntry|DeleteEntry]
) = {
```

File: ride/manager_vault.ride [hunk 2]
```diff
@@ -0,0 +1,243 @@
...
+    StringEntry(keyActiveManagerPublicKey(), pm.value().toBase58String()),
+    DeleteEntry(keyPendingManagerPublicKey())
+  ]
+}
+
+# Vote for new Manager (quorum: 2 votes)
+# Set pendingManagerPublicKey
+# Manager should confirm or activate by vote
+@Callable(i)
+func voteForNewManger(newManagerPublicKey: String) = {
```
Comment (ridev6) [inline ride/manager_vault.ride:127]:
```suggestion
func voteForNewManager(newManagerPublicKey: String) = {
```

File: ride/manager_vault.ride [hunk 3]
```diff
@@ -0,0 +1,243 @@
...
+  } else if (adminCurrentVote == 1) then {
+    throwErr(voteKey + " you already voted")
+  } else { unit }
+
+  let votes = countVotes(keyPrefix)
+  if (votes + 1 >= minVotes) then {
+    let clearVoteEntries = getClearVoteEntries(keyPrefix)
+    clearVoteEntries ++ voteResult
+  } else {
+    [ IntegerEntry(voteKey, 1)]
```
Comment (ridev6) [inline ride/manager_vault.ride:105]:
```suggestion
    [ IntegerEntry(voteKey, 1) ]
```

**Review by ridev6 — APPROVED** [review]

File: ride/manager_vault.ride [hunk 4]
```diff
@@ -0,0 +1,243 @@
...
+func keyPendingManagerPublicKey() = makeString(["%s", "pendingManagerPublicKey"], SEP)
+func keyCurrentManagerPublicKey() = makeString(["%s", "currentManagerPublicKey"], SEP)
+
+# Voting prefixes
+func keySetNewManagerVotePrefix(managerPublicKey: String) = makeString(["%s%s%s", "setManager", managerPublicKey], SEP)
+func keyDeactivateManagerVotePrefix()                     = makeString(["%s%s", "deactivateManager"], SEP)
+func keyActivateManagerVotePrefix()                       = makeString(["%s%s", "activateManager"], SEP)
+func keyAddNewAdminVotePrefix(adminCandidate: String)     = makeString(["%s%s%s", "addAdmin", adminCandidate], SEP)
+func keyRemoveAdminVotePrefix(removeCandidate: String)    = makeString(["%s%s%s", "removeAdmin", removeCandidate], SEP)
+func keyAllowedTxIdVotePrefix(txId: String)               = makeString(["%s%s%s", "allowTxId", txId], SEP)
```
Comment (epicwavesrider) [inline ride/manager_vault.ride:27]: one `%s` is unnecessary

File: ride/manager_vault.ride [hunk 5]
```diff
@@ -0,0 +1,243 @@
...
+
+func getAdminsList() = {
+  this.getString(keyAdminAddressList()).valueOrElse("").split(SEP)
+}
+
+func isInAdminList(address: String) = {
+  match (getAdminsList().indexOf(address)) {
+    case t:Int => true
+    case _ => false
+  }
```
Comment (epicwavesrider) [inline ride/manager_vault.ride:53]: you can use
```swift
containsElement(List[T], T): Boolean
```

File: ride/manager_vault.ride [hunk 6]
```diff
@@ -0,0 +1,243 @@
...
+# Admin List key
+func keyAdminAddressList() = makeString(["%s", "adminAddressList"], SEP)
+
+# Allowed TXID key
+func keyAllowedTxId() = makeString(["%s", "TXID"], SEP)
+
+func getAdminVote(prefix: String, admin: String) = {
+  let voteKey = makeString([prefix, admin], SEP)
+  getInteger(voteKey).valueOrElse(0)
+}
```
Comment (thelazarev) [inline ride/manager_vault.ride:38]: if the getAdminVote function is used in one place, as it is now, you can make the response boolean. If the key value exists, return true

File: ride/manager_vault.ride [hunk 7]
```diff
@@ -0,0 +1,243 @@
...
+
+func voteINTERNAL(
+    callerAddressString: String, 
+    keyPrefix: String, 
+    minVotes: Int, 
+    voteResult: 
+    List[StringEntry|IntegerEntry|DeleteEntry]
+  ) = {
+  let voteKey = makeString([keyPrefix, callerAddressString], SEP)
+  let adminCurrentVote = getAdminVote(keyPrefix, callerAddressString)
```
Comment (thelazarev) [inline ride/manager_vault.ride:92]: if the getAdminVote function is used in one place, as it is now, you can make the response boolean. If the key value exists, return true

File: ride/manager_vault.ride [hunk 8]
```diff
@@ -0,0 +1,243 @@
...
+
+  strict err = if (!isInAdminList(callerAddressString)) then {
+    throwErr("Address: " + callerAddressString + " not in Admin list")
+  } else if (adminCurrentVote == 1) then {
+    throwErr(voteKey + " you already voted")
+  } else { unit }
```
Comment (thelazarev) [inline ride/manager_vault.ride:98]:
```swift
let isAdminVoted = getAdminVote(keyPrefix, callerAddressString)

strict checks = [
  isInAdminList(callerAddressString) || throwErr("Address: " + callerAddressString + " not in Admin list")
  !isAdminVoted || throwErr(voteKey + " you already voted")
]
```

File: ride/manager_vault.ride (hunk 3 above)
Comment (thelazarev) [inline ride/manager_vault.ride:105]:
```suggestion
    [ IntegerEntry(voteKey, 1) ]
```

File: ride/manager_vault.ride [hunk 9]
```diff
@@ -0,0 +1,243 @@
...
+  }
+}
+
+# Vote to remove Admin address to Admin List (quorum: 2 vote)
+# Removes Admin address from adminAddressList
+@Callable(i)
+func removeAdmin(removeCandidate: String) = {
+  let callerAddressString = toBase58String(i.caller.bytes)
+  let keyPrefix = keyRemoveAdminVotePrefix(removeCandidate)
+  let quorum = 2
```
Comment (epicwavesrider) [inline ride/manager_vault.ride:193]: maybe put it in a key, and its change is also limited to multi signature?

File: ride/manager_vault.ride (hunk 8 above)
Comment (epicwavesrider) [inline ride/manager_vault.ride:98]: if the getAdminVote function is used in one place, as it is now, you can make the response boolean.
```swift
let isAdminVoted = getAdminVote(keyPrefix, callerAddressString)
strict checks = [
  isInAdminList(callerAddressString) || throwErr("Address: " + callerAddressString + " not in Admin list")
  !isAdminVoted || throwErr(voteKey + " you already voted")
]
```

File: ride/manager_vault.ride [hunk 10]
```diff
@@ -0,0 +1,243 @@
...
+    case _ => {
+      let currentAdminList = getAdminsList()
+      if (currentAdminList.size() >= MIN_ADMIN_LIST_SIZE) then {
+        (tx.id == allowedTxId)
+      } else {
+        sigVerify(tx.bodyBytes, tx.proofs[0], tx.senderPublicKey)
+      }
+    }
+  }
+}
```
Comment (epicwavesrider) [inline ride/manager_vault.ride:243]: add newline please

File: ride/manager_vault.ride [hunk 11]
```diff
@@ -0,0 +1,243 @@
...
+  voteINTERNAL(callerAddressString, keyPrefix, quorum, result)
+}
+
+# Vote to Add new Admin address to Admin List (quorum: 2 vote)
+# Appends new address to adminAddressList
+@Callable(i)
+func addNewAdmin(newAdminAddress: String) = {
+  let callerAddressString = toBase58String(i.caller.bytes)
+  let keyPrefix = keyAddNewAdminVotePrefix(newAdminAddress)
+  let quorum = 2
```
Comment (epicwavesrider) [inline ride/manager_vault.ride:167]: maybe put it in a key, and its change is also limited to multi signature?

File: ride/manager_vault.ride [hunk 12]
```diff
@@ -0,0 +1,243 @@
...
+  voteINTERNAL(callerAddressString, keyPrefix, quorum, result)
+}
+
+# Vote to Activate Current Manager (quorum: 2 vote)
+# Copies pendingManagerPublicKey to managerPublicKey
+@Callable(i)
+func activateManager() = {
+  let callerAddressString = toBase58String(i.caller.bytes)
+  let keyPrefix = keyActivateManagerVotePrefix()
+  let quorum = 2
```
Comment (epicwavesrider) [inline ride/manager_vault.ride:154]: maybe put it in a key, and its change is also limited to multi signature?

File: ride/manager_vault.ride [hunk 13]
```diff
@@ -0,0 +1,243 @@
...
+  voteINTERNAL(callerAddressString, keyPrefix, quorum, result)
+}
+
+# Vote to Deactivate Manager (quorum: 1 vote)
+# Clears managerPublicKey value
+@Callable(i)
+func deactivateManager() = {
+  let callerAddressString = toBase58String(i.caller.bytes)
+  let keyPrefix = keyDeactivateManagerVotePrefix()
+  let quorum = 1
```
Comment (epicwavesrider) [inline ride/manager_vault.ride:142]: maybe put it in a key, and its change is also limited to multi signature?

File: ride/manager_vault.ride [hunk 14]
```diff
@@ -0,0 +1,243 @@
...
+}
+
+# Vote for new Manager (quorum: 2 votes)
+# Set pendingManagerPublicKey
+# Manager should confirm or activate by vote
+@Callable(i)
+func voteForNewManger(newManagerPublicKey: String) = {
+  let callerAddressString = toBase58String(i.caller.bytes)
+  let keyPrefix = keySetNewManagerVotePrefix(newManagerPublicKey)
+  let quorum = 2
```
Comment (epicwavesrider) [inline ride/manager_vault.ride:130]: maybe put it in a key, and its change is also limited to multi signature?

File: ride/manager_vault.ride (hunk 6 above)
Comment (epicwavesrider) [inline ride/manager_vault.ride:38]: if the getAdminVote function is used in one place, as it is now, you can make the response boolean

File: ride/manager_vault.ride (hunk 2 above)
Comment (bra1nsurfer) [reply ride/manager_vault.ride:127]: Fixed
Updated in tests

File: ride/manager_vault.ride (hunk 4 above)
Comment (bra1nsurfer) [reply ride/manager_vault.ride:27]: Prefixes is later concatenated with admin address to make full vote key
`"%s%s__activateManager" + "__" + "{adminAddress}"`
Added explicit `keyFullAdminVote(prefix: String, adminAddress: String)` function

File: ride/manager_vault.ride (hunk 5 above)
Comment (bra1nsurfer) [reply ride/manager_vault.ride:53]: Swapped to `containsElement`

File: ride/manager_vault.ride (hunk 10 above)
Comment (bra1nsurfer) [reply ride/manager_vault.ride:243]: Fixed

File: ride/manager_vault.ride (hunk 6 above)
Comment (bra1nsurfer) [reply ride/manager_vault.ride:38]: It was made integer to simplify quorum calculations in `countVotes()`

File: test/components/manager_vault/_hooks.mjs [hunk 15]
```diff
@@ -0,0 +1,58 @@
...
+    const massTransferTx = massTransfer({
+      transfers: seeds.map((item) => ({ recipient: address(item, chainId), amount })),
+      chainId,
+    }, seed);
+    await api.transactions.broadcast(massTransferTx, {});
+    await waitForTx(massTransferTx.id, { apiBase });
+
+    await setScriptFromFile(managerVaultPath, this.accounts.managerVault);
+
+    const adminsListString = `${address(this.accounts.admin1, chainId)}__${address(this.accounts.admin2, chainId)}__${address(this.accounts.admin3, chainId)}`;
```
Comment (ridev6) [inline test/components/manager_vault/_hooks.mjs:43]: `join` can be used

File: test/components/manager_vault/_hooks.mjs [hunk 16]
```diff
@@ -0,0 +1,58 @@
...
+  data,
+  massTransfer,
+  nodeInteraction,
+} from '@waves/waves-transactions';
+import { create } from '@waves/node-api-js';
+import { format } from 'path';
+import { setScriptFromFile } from '../../utils/utils.mjs';
+
+const { waitForTx } = nodeInteraction;
+const apiBase = process.env.API_NODE_URL;
```
Comment (ridev6) [inline test/components/manager_vault/_hooks.mjs:12]: can be imported from [api.mjs](https://github.com/waves-exchange/contracts/blob/main/test/utils/api.mjs#L4)

**Review by ridev6 — CHANGES_REQUESTED** [review]

File: test/components/manager_vault/_hooks.mjs (hunk 15 above)
Comment (bra1nsurfer) [reply test/components/manager_vault/_hooks.mjs:43]: fixed

File: test/components/manager_vault/_hooks.mjs (hunk 16 above)
Comment (bra1nsurfer) [reply test/components/manager_vault/_hooks.mjs]: Changed in all tests

**Review by ridev6 — APPROVED** [review]
LGTM

**Pull request merged**